integration/flink/app/build.gradle

```diff
@@ -0,0 +1,368 @@
+        outputDir.deleteDir()
+    }
+}
+
+javadoc {
+    dependsOn delombok
+    source = delombok.outputDir
+    options.tags = [ "apiNote" ]
+}
+
+spotless {
+    def disallowWildcardImports = {
+        String text = it
+        def regex = ~/import .*\.\*;/
+        def m = regex.matcher(text)
+        if (m.find()) {
+            throw new AssertionError("Wildcard imports disallowed - ${m.findAll()}")
+        }
+    }
+    java {
+        googleJavaFormat()
+        removeUnusedImports()
+        custom 'disallowWildcardImports', disallowWildcardImports
+    }
+}
+
+def reportsDir = "${buildDir}/reports"
+def coverageDir = "${reportsDir}/coverage"
+
+jacoco {
+    toolVersion = '0.8.5'
+    reportsDir = file(coverageDir)
+}
+
+jacocoTestReport {
+    reports {
+        xml {
+            enabled true
+        }
+        html {
+            enabled true
+            destination = file(coverageDir)
+        }
+    }
+}
+
+publishing {
+    publications {
+        mavenJava(MavenPublication) {
+            groupId = 'io.openlineage'
+            artifactId = 'openlineage-flink'
+
+            from components.java
+
+            artifact sourceJar
+            artifact javadocJar
+
+            pom {
+                name = 'openlineage-flink'
+                description = 'OpenLineage integration with Flink'
+                url = 'https://github.com/OpenLineage/OpenLineage'
+                licenses {
+                    license {
+                        name = 'The Apache License, Version 2.0'
+                        url = 'http://www.apache.org/licenses/LICENSE-2.0.txt'
+                    }
+                }
+                developers {
+                    developer {
+                        id = 'openlineage'
+                        name = 'OpenLineage Project'
+                    }
+                }
+                scm {
+                    connection = 'scm:git:git://github.com/OpenLineage/OpenLineage.git'
+                    developerConnection = 'scm:git:ssh://github.com:OpenLineage/OpenLineage.git'
+                    url = 'https://github.com/OpenLineage/OpenLineage'
+                }
+            }
+        }
+    }
+
+    processResources {
+        filter ReplaceTokens, tokens: [
+                "version": project.property("version")
+        ]
+    }
+
+    repositories {
+        maven {
+            url = 'https://oss.sonatype.org/service/local/staging/deploy/maven2'
+            credentials {
+                username = System.getenv('RELEASE_USERNAME')
+                password = System.getenv('RELEASE_PASSWORD')
+            }
+        }
+    }
+}
+
+signing {
+    required { isReleaseVersion }
+    def signingKey = findProperty("signingKey")
+    def signingPassword = findProperty("signingPassword")
+    useInMemoryPgpKeys(signingKey, signingPassword)
+    sign publishing.publications.mavenJava
+}
+
+shadowJar {
+    dependsOn(':examples:stateful:shadowJar')
+    dependsOn(':examples:stateful:pmdMain')
```

Comment:
Please, do not depend on the tasks of other modules. This can cause Gradle to perform unnecessary configuration work. If you need the outputs of the ":examples:stateful:shadowJar" task, then you should declare a dependency on the `:examples:stateful` module like so:

```groovy
runtimeOnly(project(path: ':examples:stateful', configuration: 'shadow'))
```

The dependency on `:examples:stateful:pmdMain` is confusing to me though. Could you explain why you depend on another modules `pmdMain` task?

Reply:
I moved the original build.gradle from root to the app module. I think it just to make integration test run after all of the shadow jar are created. 

https://github.com/OpenLineage/OpenLineage/blob/main/integration/flink/build.gradle#L306

After changing the to runtimeOnly(project(path: ':examples:stateful', configuration: 'shadow')). Error comes from build.

Could not find method project() for arguments [{path=:examples:stateful, configuration=shadow}] on task ':app:shadowJar' of type com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar

Reply:
Oh, my apologies, I should have been more clearer.

The dependency should be added to the module's `dependencies { }` block.

Concretely,

```groovy
dependencies {
	// other dependencies

	runtimeOnly(project(path: ':examples:stateful', configuration: 'shadow'))

	// test dependencies
}
```

You can then verify that the project appears on the runtime classpath by doing the following:

```groovy
tasks.register("printSourceSetDependencyInformation") {
    group = "debug"
    description = "Prints the dependencies per configuration for a given source set, use '-Psource.set.name=<value>'"

    doLast {
        final String sourceSetName = project.findProperty('source.set.name').toString()
        if (sourceSetName == null) {
            throw new GradleException("Please provide a source set name using '-Psource.set.name=<value>'")
        }

        final SourceSet sourceSet = sourceSets.getByName(sourceSetName)
        if (sourceSet == null) {
            throw new GradleException("Source set with name $sourceSetName does not exist.")
        }

        final List<String> configurationNames = Arrays.stream(SourceSet.class.getDeclaredMethods())
                .filter { Method m -> m.name.startsWith("get") && m.name.endsWith("ConfigurationName") }
                .map { Method m -> m.invoke(sourceSet).toString() }
                .sorted(Comparator.naturalOrder())
                .collect(Collectors.toList())

        Consumer<String> printer = { String configurationName ->
            println("Dependencies for configuration: ${configurationName}")

            try {
                final def deps = configurations.getByName(configurationName).dependencies
                if (deps.isEmpty()) {
                    println("No dependencies have been added to this configuration.")
                } else {
                    deps.forEach { Dependency dependency ->
                        println("- ${dependency.group}:${dependency.name}:${dependency.version}")
                    }
                }
            } catch (UnknownConfigurationException e) {
                println("Configuration $configurationName does not exist.")
                return
            } finally {
                println()
            }
        }

        configurationNames.forEach { String configurationName ->
            printer.accept(configurationName)
        }
    }
}
```

You can run the above task like so:

```bash
./gradlew app:printSourceSetDependencyInformation -Psource.set.name=main
```

You want to make sure that the `runtimeClasspath` has the `example stateful` entry there. (The axact name can very, based on how the projects are configured)

Reply:
@d-m-h 
Thanks for the suggestion. Using the fixtures jar in dependencies is a better solution. It does't need to build shadow jar for app module. Please review the change.